kernel/src/main/java/org/kframework/compile/checks/CheckAtt.java

```diff
@@ -3,28 +3,57 @@
 
 import org.kframework.attributes.Att;
 import org.kframework.attributes.HasLocation;
+import org.kframework.builtin.Sorts;
+import org.kframework.compile.ExpandMacros;
 import org.kframework.definition.Module;
+import org.kframework.definition.Production;
 import org.kframework.definition.Rule;
 import org.kframework.definition.Sentence;
+import org.kframework.kore.KLabel;
 import org.kframework.utils.errorsystem.KEMException;
 
 import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.kframework.Collections.*;
 
 /**
  * Created by dwightguth on 1/25/16.
  */
 public class CheckAtt {
+    private final Set<KLabel> macros;
     private final Set<KEMException> errors;
     private final Module m;
+    private final boolean isSymbolicKast;
 
-    public CheckAtt(Set<KEMException> errors, Module m) {
+    public CheckAtt(Set<KEMException> errors, Module m, boolean isSymbolicKast) {
         this.errors = errors;
         this.m = m;
+        this.isSymbolicKast = isSymbolicKast;
+        this.macros = stream(m.rulesFor()).filter(e -> stream(e._2()).filter(r -> ExpandMacros.isMacro(r)).findAny().isPresent()).map(e -> e._1()).collect(Collectors.toSet());
     }
 
     public void check(Sentence sentence) {
         if (sentence instanceof Rule) {
             check(((Rule) sentence).att(), sentence);
+        } else if (sentence instanceof Production) {
+            check((Production) sentence);
+        }
+    }
+
+    private void check(Production prod) {
+        if (!prod.sort().equals(Sorts.KItem())) {
+            Att sortAtt =  m.sortAttributesFor().getOrElse(prod.sort().head(), () -> Att.empty());
+            if (sortAtt.contains(Att.HOOK()) && !sortAtt.get(Att.HOOK()).equals("ARRAY.Array") && !(sortAtt.get(Att.HOOK()).equals("KVAR.KVar") && isSymbolicKast)) {
+                if (!prod.att().contains(Att.FUNCTION()) && !prod.att().contains(Att.BRACKET()) &&
+                    !prod.att().contains("token") && !(prod.klabel().isDefined() && macros.contains(prod.klabel().get()))) {
+                    if (!(prod.sort().equals(Sorts.K()) && ((prod.klabel().isDefined() && (prod.klabel().get().name().equals("#EmptyK") || prod.klabel().get().name().equals("#KSequence"))) || prod.isSubsort()))) {
+                        if (!(sortAtt.contains("cellCollection") && prod.isSubsort())) {
+                            errors.add(KEMException.compilerError("Cannot add new constructors to hooked sort " + prod.sort(), prod));
+                        }
+                    }
+                }
+            }
```

Comment:
Can we instead have a pass which runs before this and does:

-  The same check, and if it passes adds the `constructor` attribute if it determines that the production is an attribute.
-  Then this check is changed to loop over all productions with `constructor` attribute and make sure none of them are targeting builtin sorts.

This has the benefit of (i) clearly defining what we call a "constructor", and (ii) adding the attribute to productions in case it's useful to the other various tools/backends.

Reply:
There are several problems with this approach:

1. The checks normally happen before any transformation happens to the k definition. We might be able to change this, but it would be a substantial architectural change and I'd rather not make such a fundamental change on a completely ad-hoc basis. If we're going to do something like this, it would be better to have an explicit semantic analysis phase and lay oiut exactly how it should work and convey information.
2. Transformers that only add attributes to sentences don't really work. I've tried repeatedly to try to bypass this restriction and it's never gone well. It's well outside the scope of this change.
3. Your proposed solution would disallow sentences that aren't constructors that the user put the constructor attribute on in built-in sorts, which would be a breaking change (albeit one probably not exercised by anything). But it would be very confusing.